RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitProducerKernel.dev.cc

```diff
@@ -0,0 +1,199 @@
+    pfrh.detId() = rh.detId();
+    pfrh.energy() = rh.energy();
+    pfrh.time() = rh.time();
+    pfrh.depth() = HCAL::getDepth(pfrh.detId());
+    const uint32_t subdet = getSubdet(pfrh.detId());
+    if (subdet == HcalBarrel)
+      pfrh.layer() = PFLayer::HCAL_BARREL1;
+    else if (subdet == HcalEndcap)
+      pfrh.layer() = PFLayer::HCAL_ENDCAP;
+    else
+      pfrh.layer() = PFLayer::NONE;
+  }
+
+  template <>
+  ALPAKA_FN_ACC void PFRecHitProducerKernelImpl1<ECAL>::ConstructPFRecHit(
+      PFRecHitDeviceCollection::View::element pfrh, const ECAL::CaloRecHitSoATypeDevice::ConstView::const_element rh) {
+    pfrh.detId() = rh.detId();
+    pfrh.energy() = rh.energy();
+    pfrh.time() = rh.time();
+    pfrh.depth() = 1;
+    const uint32_t subdet = getSubdet(pfrh.detId());
+    if (subdet == EcalBarrel)
+      pfrh.layer() = PFLayer::ECAL_BARREL;
+    else if (subdet == EcalEndcap)
+      pfrh.layer() = PFLayer::ECAL_ENDCAP;
+    else
+      pfrh.layer() = PFLayer::NONE;
+  }
+
+  template <typename CAL>
+  struct PFRecHitProducerKernelImpl2 {
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const typename CAL::TopologyTypeDevice::ConstView topology,
+                                  PFRecHitDeviceCollection::View pfRecHits,
+                                  const uint32_t* __restrict__ denseId2pfRecHit,
+                                  uint32_t* __restrict__ num_pfRecHits) const {
+      // First thread updates size field pfRecHits SoA
+      if (const int32_t thread = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc)[0u]; thread == 0)
+        pfRecHits.size() = *num_pfRecHits;
+
+      // Assign position information and associate neighbours
+      for (int32_t i : cms::alpakatools::elements_with_stride(acc, *num_pfRecHits)) {
+        const uint32_t denseId = CAL::detId2denseId(pfRecHits.detId(i));
+
+        pfRecHits.x(i) = topology.positionX(denseId);
+        pfRecHits.y(i) = topology.positionY(denseId);
+        pfRecHits.z(i) = topology.positionZ(denseId);
+
+        for (uint32_t n = 0; n < 8; n++) {
+          pfRecHits.neighbours(i)(n) = -1;
+          const uint32_t denseId_neighbour = topology.neighbours(denseId)(n);
+          if (denseId_neighbour != 0xffffffff) {
+            const uint32_t pfRecHit_neighbour = denseId2pfRecHit[denseId_neighbour];
+            if (pfRecHit_neighbour != 0xffffffff)
+              pfRecHits.neighbours(i)(n) = (int32_t)pfRecHit_neighbour;
+          }
+        }
+      }
+    }
+  };
+
+  template <typename CAL>
+  PFRecHitProducerKernel<CAL>::PFRecHitProducerKernel(Queue& queue)
+      : denseId2pfRecHit_(cms::alpakatools::make_device_buffer<uint32_t[]>(queue, CAL::SIZE)),
+        num_pfRecHits_(cms::alpakatools::make_device_buffer<uint32_t>(queue)),
```

Comment:
Yes, they hold temporary event-specific data. I was trying to avoid frequent memory allocations, by keeping these buffers between events. During my testing, there was no notable performance difference though.
When I go with your suggestion, the `kernel` object in `PFRecHitProducerAlpaka::produce` can be made a local variable. This also means that the module can be made a global producer. If the kernel object is created separately for each event anyway, the buffer allocations might as well be kept in the constructor.

Reply:
With the new benchmarking method (running the producer multiple times per event), I can measure the performance impact of this issue. Going from the current version (memory buffer lives forever) to your suggestion (allocate buffer for each event) increases the runtime on the GPU backend from ~9us/event to ~16us/event (for HCAL). This is negligible compared to the total HLT time, so I think it is better to stick with your suggestion and have cleaner code.